lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+              e.twin.next = e
+              e.prev = e.twin
+            end
+          end
+        end
+
+        # It is possible that intersections occur when new edges are added.
+        # This will split those edges into more half-edges while preserving
+        # the existing half-edges when possible since geometries may reference
+        # them.
+        def compute_split_edges(seg, ints)
+          points = ints.uniq.sort do |a, b|
+            a.distance(seg.s) <=> b.distance(seg.s)
+          end
+          points = [seg.s] + points + [seg.e]
+
+          he_start = incident_edges[points.first.coordinates].find { |he| he.destination == points.last }
+          he_end = incident_edges[points.last.coordinates].find { |he| he.destination == points.first }
+
+          points.each_cons(2) do |s, e|
+            edge = Segment.new(s, e)
+            if s == he_start.origin
+              he = HalfEdge.new(e)
+              he_start.twin = he
+              he.twin = he_start
+              insert_half_edge(he)
+            elsif e == he_end.origin
+              he = HalfEdge.new(s)
+              he_end.twin = he
+              he.twin = he_end
+              insert_half_edge(he)
+            else
+              create_half_edge(edge)
+            end
+            @edges << edge
+          end
+          @edges.delete(seg)
+        end
+      end
+
+      # GeometryGraph is a PlanarGraph that is built by adding
+      # geometries instead of edges. The GeometryGraph will
+      # hold a reference to an arbitrary HalfEdge on the
+      # interior of the geometry for every boundary in the geometry.
+      # For example, a polygon will have a reference to a HalfEdge for its
+      # exterior shell and one for every hole.
+      class GeometryGraph < PlanarGraph
+        # GeomEdge will be used to store the references to the HalfEdges
+        GeomEdge = Struct.new(:exterior_edge, :interior_edges)
+
+        def initialize(geom)
+          super()
+          @parent_geometry = geom
+          @geom_edges = []
+          add_geometry(geom)
+        end
+        attr_reader :parent_geometry, :geom_edges
+
+        private
+
+        # Adds a geometry to the graph and finds its
+        # reference HalfEdge(s).
+        #
+        # @param geom [RGeo::Feature::Instance]
+        def add_geometry(geom)
+          case geom
+          when Feature::Point
+            # Can't handle points yet, so just add an empty entry for them
+            @geom_edges << GeomEdge.new
+          when Feature::LineString, Feature::LinearRing
+            add_line_string(geom)
+          when Feature::Polygon
+            add_polygon(geom)
+          when Feature::GeometryCollection
+            add_collection(geom)
+          end
+        end
+
+        # Adds a LineString or LinearRing
+        # to the graph.
+        #
+        # @param geom [RGeo::Feature::LineString]
+        def add_line_string(geom)
+          # TODO: should we handle empty linestrings?
```

Comment:
can we handle those smoothly, or do you talk about raising ?

Reply:
We should probably treat those the same way as points.

Reply:
Added a test, but this is handled smoothly.

Reply:
then can't we get read of the TODO? :)